### PR TITLE
Feature/gptcachelitellm

### PR DESCRIPTION
**Pull request by manikesh12**
#### Description

This PR implements gptcache caching method in litellm.

#### Methodology/Tasks Implemented

- implement setting the cache and lookup for data.
-implement asynchronous versions of them

#### Testing

- Test with test_caching.py in test folder in litellm.

#### Estimated Time

- Approximately 10 hours.

### CODE REVIEW

File: litellm/caching.py [hunk 1]
```diff
@@ -25,7 +25,8 @@
 from litellm.litellm_core_utils.core_helpers import _get_parent_otel_span_from_kwargs
 from litellm.types.services import ServiceLoggerPayload, ServiceTypes
 from litellm.types.utils import all_litellm_params
-
+import os
```
Comment (sonalibud) [inline litellm/caching.py]: This change is not required for gptcache faeture. Please remove.

File: litellm/caching.py [hunk 2]
```diff
@@ -1669,13 +1692,14 @@ def delete_cache(self, key):
...
 
+from .redis_gpt_cache import RedisGPTCache
 
 #### LiteLLM.Completion / Embedding Cache ####
 class Cache:
     def __init__(
         self,
         type: Optional[
-            Literal["local", "redis", "redis-semantic", "s3", "disk"]
+            Literal["local", "redis", "redis-semantic", "redis-gptcache", "s3", "disk"]
```
Comment (sonalibud) [inline litellm/caching.py]: Let this name be "gpt_cache_redis". This is what we used in budserve_app. Consistency will be maintained.

File: litellm/caching.py [hunk 3]
```diff
@@ -930,22 +930,45 @@ def __init__(
             "index": {
                 "name": "litellm_semantic_cache_index",
                 "prefix": "litellm",
-                "storage_type": "hash",
+                # "key_separator": ":",
```
Comment (sonalibud) [inline litellm/caching.py]: Are these cahnges required ? These are in RedisSemanticCache class, isn't it ? We do not need to modify anything in RedisSemanticCache class.

File: litellm/caching.py [hunk 4]
```diff
@@ -1721,9 +1745,12 @@ def __init__(
         s3_config: Optional[Any] = None,
         s3_path: Optional[str] = None,
         redis_semantic_cache_use_async=False,
+        redis_gptcache_use_async=False,
```
Comment (sonalibud) [inline litellm/caching.py]: no need to add extra flags for gptcache, you can use value of redis_semantic cache flags in order to specify use_async behaviour or define embedding model. If you add more flags, then their mapping at every place will also be required.

File: litellm/caching.py [hunk 5]
```diff
@@ -1721,9 +1745,12 @@ def __init__(
...
         redis_semantic_cache_embedding_model="text-embedding-ada-002",
+        redis_gptcache_embedding_model="sentence-transformers/all-MiniLM-L6-v2",
         redis_flush_size=None,
         disk_cache_dir=None,
+
```
Comment (sonalibud) [inline litellm/caching.py]: remove this extra line

**Review by sonalibud — CHANGES_REQUESTED** [review]

File: litellm/redis_gpt_cache.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        print_verbose(f"gptcache redis semantic-cache using Onnx embeddings")
+        embeddings = Onnx()
+    eviction_policy = cache_config.get("eviction_policy", {})
+
+    data_manager = manager_factory(
+        "redis,redis",
+        scalar_params={
+            "redis_host": os.getenv("REDIS_HOST", "localhost"),
+            "redis_port": os.getenv("REDIS_PORT", "6342"),
+            "password": os.getenv("REDIS_PASSWORD", "budpassword"),
```
Comment (sonalibud) [inline litellm/redis_gpt_cache.py]: It's not alright to keep password values here. Please check how redis connection params are being set in RedisSemanticCache. I had kept it only for testing, but when we set up proxy server using config file, these values should be passed on from there. Please correct that.

File: litellm/redis_gpt_cache.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+        "policy": eviction_policy.get("policy", "LRU"),
+        "clean_size": int(eviction_policy.get("max_size", 100) * 0.2) or 1,
+        "ttl": eviction_policy.get("ttl"),
+        "on_evict": data_manager._clear,
+    }
+    data_manager.eviction_base = BudServeMemoryCacheEviction(**eviction_params)
+
+    ids = data_manager.s.get_ids(deleted=False)
+    data_manager.eviction_base.put(ids)
+    from gptcache.processor.pre import get_prompt
```
Comment (sonalibud) [inline litellm/redis_gpt_cache.py]: Why is this line here ? Where is it being used ?

**Review by sonalibud — CHANGES_REQUESTED** [review]

**Comment by sonalibud** [comment]
Reviewed. Integration testing required with litellm global proxy auth changes.

File: litellm/redis_gpt_cache.py [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+                port = os.getenv("REDIS_PORT")
+                password = os.getenv("REDIS_PASSWORD")
+                if host is None or port is None or password is None:
+                    raise Exception("Redis host, port, and password must be provided")
+
+            redis_url = "redis://:" + password + "@" + host + ":" + port
+        self.redis_params = {"host": host, "port": port, "password": password}
+        GPTCache.__init__(self, init_gptcache_redis)
+        print_verbose(f"gptcache redis semantic-cache redis_url: {redis_url}")
+        self.redis_params = {"host": host, "port": port, "password": password}
```
Comment (sonalibud) [inline litellm/redis_gpt_cache.py]: redis_params and GPTCache are initialized twice, please correct

**Review by sonalibud — CHANGES_REQUESTED** [review]

File: litellm/redis_gpt_cache.py (hunk 8 above)
Comment (manikesh12) [reply litellm/redis_gpt_cache.py]: noted
